Add assets loading, binary selection and VCS information in manifest

krnowak opened the pull request:
This PR fixes #4, #6 and #7. It also refactors a code which might look like a rewrite. Please see the issues mentioned to see how fixes were implemented.

krnowak committed:
Put options into separate function

Just a tidbit of refactoring.

krnowak committed:
Allow specifying which binary to put into ACI

Add --use-binary option which allows user to choose which binary
should be placed in the image if go get builds multiple binaries.

This change also introduces some changes in naming - when project
example.com/foo/bar has binaries bar and quux, then for quux the
created image will have a filename bar-quux.aci, name in manifest will
be example.com/foo/bar-quux.

krnowak committed:
Allow specifying assets

Add --asset option which takes two paths separated with OS specific
path separator (colon on Unices, semicolon on Windows). The first path
specifies a path inside produced image, second part - the path from
local filesystem which will be copied into image.

We handle only regular files, directories and symlinks that does not
point outside the asset.

krnowak committed:
Allow preserving temporary directory

This is strictly developer option to be able to view the contents of
/tmp/goaciXXX directory, when something goes wrong.

krnowak committed:
Add VCS specific label to manifest

It supports getting information from git, hg, svn and bzr.

krnowak committed:
Move asset code into separate file

The goaci.go already looks like a kitchen sink with two year old
leftovers. Let's clean it up a bit.

krnowak committed:
Preserve absolute symlink targets

If a symlink target is an absolute path still pointing somewhere
inside the asset it sits in then preserve the absoluteness even in
rootfs. That requires a bit of path munging.

krnowak committed:
Add placeholders for assets

Add two placeholders - GOPATH and PROJPATH. GOPATH placeholder is
obvious. PROJPATH placeholder will be replaced with absolute path to
project go get is building. Usage is simple:

--asset /assets:<PROJPATH>/assets

krnowak committed:
Refactor asset handling

Split out some code to separate functions with hopefully
self-explanatory names. The additional aim is to have function fit
into single screen.

krnowak committed:
Put misc functions in separate file

Remove some more clutter from goaci.go.

krnowak committed:
More reliable project option parsing

We are using flag package for option parsing, so stop using os.Args.
We can take the project from flag.Args after ensuring that it has one
and only one string.

krnowak committed:
The grand refactoring of code

This basically rewrites goaci.go into many smaller and clearer
functions. It also removes all uses of die function, so the deferred
removal of /tmp/goaciXXX is actually executed (die used os.Exit()
which prevented that).

krnowak committed:
Add documentation for some more obscure functions.

krnowak committed:
Update README

alban commented:
@iaguis : you said out of band that you had a problem with using goaci on etcd with this branch? Is it still the case?

jonboulle commented:
@alban @iaguis could you guys please review this if you haven't already?

alban commented on goaci.go:
```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+// goaci.aci. For project with multiple binaries
+// (github.com/appc/spec/...) returned values would be (assuming ace
+// as selected binary): github.com/appc/spec, github.com/appc/spec-ace
+// and spec-ace.aci.
```
I am not convinced by the need to change the ACName and the ACI filename. I don't think the user would want to create several ACIs with different names and each pointing to a different binary. Rather, they would produce one ACI containing all the binaries and starting the main binary selected by --use-binary.

The current etcd ACI (built without `goaci`) contains `etcd`, `etcdctl` and documentation files. The documentation can be added with --asset but including several binaries in a single ACI is useful in general: the main binary could fork and execute the other binaries. (I don't think etcd executes etcdctl but other projects could do that)

I see that `etcdctl` is not build by `go build`. It is only built by `./build`. Does it means there is no way to include both etcd and etcdctl in the same ACI? I wonder if there is a way to reorganise the etcd sources so that `go build` would produce both binaries and to include them both in the ACI.

krnowak replied:
My impression is that ACI image is basically a package of one standalone application and its assets, in this case - etcd daemon. If you want more binaries, create more ACI images. That's why I'm appending binary name to project name if it produces more than one binary. The case would be even worse if you have project `foo` that builds two daemons `bar` and `baz`. I suppose it would be better to create differently named images for daemon `bar` (like `foo-bar.aci`) and for daemon `baz` (like `foo-baz.aci`) instead of having `foo.aci` and wondering if it runs `bar` or `baz` daemon.

If a binary forks and executes other binaries then it is builder's responsibility to add them as an asset.

As of `./build` - sounds like user has to run `goaci github.com/coreos/etcd/...` to get both `etcd` and `etcdctl` binary. I haven't tested that.

alban commented on asset.go:
```diff
@@ -0,0 +1,173 @@
+
+// copySymlink copies the symlink, but before doing so it ensures that
+// the symlink is pointing to node inside the asset.
+//
```
Why is it necessary to check whether the symlink point outside the asset? You are creating the symlinks without dereferencing them, so there is no risks to copy something which does not belong to the asset in the ACI.

It prevents the user from creating a symlink to the compiled binary (which is not an asset) or to a file in a different asset (when several assets are given in parameters). I'm thinking of projects like Busybox with one binary and several symlinks to the same binary. The symlink `$HOME/some/asset/mount` pointing to `/busybox` would be broken on the build machine (let's say busybox is not installed on the build machine, or at least not at this path...) but once the ACI is built, it would not be broken in the container (because `/busybox` would be provided by the compilation).

alban replied:
In general, I think it is the responsibility of the person preparing the assets to make sure the symlinks link to a path which makes sense once it is copied in the ACI. That would make the code in goaci a lot easier, since it would not have to try to understand whether the symlink is relative or absolute, inside the asset or outside the asset.

krnowak replied:
Hm, could be. Maybe instead of mangling symlink targets and whatnot I should just add a code looking for dangling symlinks after all assets were added and warn about them.

alban replied:
It would look good to me.

And the check for dangling symlinks could be done later, it does not need to be part of this PR.

krnowak replied:
Right. It's get complicated a bit. If we will want to check for dangling symlinks then some symlink target mangling code will be needed anyway. Unless we get some chroot superpowers as non-root.

alban commented on asset.go:
```diff
@@ -0,0 +1,173 @@
+			return err
+		}
+	} else {
+		return fmt.Errorf("Symlink %s points to %s, which is outside asset %s", src, symTarget, root)
```
You can use %q for [quoting strings](https://golang.org/pkg/fmt/).

krnowak replied:
Thanks, useful for nasty filenames with spaces.

alban commented on asset.go:
```diff
@@ -0,0 +1,173 @@
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return err
+	}
```
defer close(srcFile)?

krnowak replied:
Thanks for catching it.

krnowak replied:
And BTW. `srcFile.Close()` ;)

alban commented on asset.go:
```diff
@@ -0,0 +1,173 @@
+	destFile, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
```
defer close(destFile)?

krnowak replied:
Thanks for catching it.

alban commented on goaci.go:
```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+func getVCSLabel(pathsNames *pathsAndNames) (*types.Label, error) {
+	name, value, err := GetVCSInfo(pathsNames.projectPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get VCS info: %v", err)
```
Instead of failing to build the aci, what about printing a warning and continue without the label?

Rationale: I would like if in the future it is possible to download a tarball of a etcd release (which does not contain the VCS info), do a couple of changes, and build the ACI with `goaci /tmp/etcd-v2.0.5-linux-amd64`.

krnowak replied:
I'm not sure I'd like to add support for that. That would probably need adding another option telling whether last parameter passed to goaci is a project name or a directory which it should try work on. It can get ambiguous in one corner case - you are in $GOPATH/src and call `goaci github.com/coreos/etcd`. Should `goaci` try to basically do `cd github.com/coreos/etcd` or rather run `go get github.com/coreos/etcd`?

But I'd rather file a separate issue about it.

alban replied:
Good point. Fair enough, a separate issue is ok to me.

alban commented on util.go:
```diff
@@ -0,0 +1,53 @@
+	fmt.Fprintln(w, strings.TrimSuffix(s, "\n"))
+}
+
+func Warn(i ...interface{}) {
```
Since Warn() is only used from "package main", no need to export it with a capital "W".

Same thing for other functions below.

krnowak replied:
I prefer to keep it that way. Later I may want to make goaci a library and this clearly says which functions are going to be exported.

alban commented on vcs.go:
```diff
@@ -0,0 +1,118 @@
+
+type VCSInfo interface {
+	IsValid(path string) bool
+	GetLabelAndId(path string) (string, string, error)
```
Awesome tests. Sad that the [vcs package](https://godoc.org/golang.org/x/tools/go/vcs) does not provide what we need here.

The sources of the vcs package are on https://github.com/golang/tools/tree/master/go/vcs but the issues are tracked on the [golang/go tracker](https://github.com/golang/go/issues).

Can you open an issue on the vcs package requesting the missing features and add a comment here in goaci with a link to the issue?

krnowak replied:
I've seen the code there, but I couldn't find out if it is exported. Thanks.

If we could use that then IsValid is not needed - we could just use [RepoRootForImportPath](https://godoc.org/golang.org/x/tools/go/vcs#RepoRootForImportPath) or [FromDir](https://godoc.org/golang.org/x/tools/go/vcs#FromDir).

Issue filed - https://github.com/golang/go/issues/10159

alban commented on goaci.go:
```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+}
+
+func copyAssets(opts *options, pathsNames *pathsAndNames) error {
+	paths := map[string]string{
```
Instead of `paths`, I would name it `placeholderMapping` or something more descriptive.

Same thing in PrepareAssets() below.

krnowak replied:
Done.

alban commented on goaci.go:
```diff
@@ -47,189 +27,398 @@ func (v *StringVector) Set(str string) error {
+	defer tr.Close()
+
+	iw := aci.NewImageWriter(*manifest, tr)
+	if err := filepath.Walk(pathsNames.aciDirPath, aci.BuildWalker(pathsNames.aciDirPath, iw)); err != nil {
```
Can we add a comment such as the following?

// FIXME: the files in the tar archive are added with the wrong uid/gid.
// The uid/gid of the aci builder leaks in the tar archive.
// See: https://github.com/appc/goaci/issues/16

krnowak replied:
Done.

alban commented on asset.go:
```diff
@@ -0,0 +1,173 @@
+	return fmt.Errorf("Can't handle local asset %v - not a file, not a dir", fi.Name())
+}
+
+func PrepareAssets(assets []string, rootfs string, paths map[string]string) error {
```
No need to export PrepareAssets with a capital P since it is not used outside of package main.

krnowak replied:
Same case in Warn function case - I prefer to leave for future librarification.

iaguis commented:
Great review, @alban. I looked into it as well and I couldn't find anything else.

krnowak commented:
Thanks for review @alban, the fixes will come with followup commits.

krnowak committed:
Lift restrictions for symlinks in assets

That simplifies the code a lot for a short while. Later some of it
will probably come back when we implement checking for dangling
symlinks after assets are copied.

krnowak committed:
Put some strings into quotes

These usually are filenames, so it's good to delimit them with quotes
in case there are spaces in them.

krnowak committed:
Close files after copying asset

krnowak committed:
Use better variable name

krnowak committed:
Add a FIXME for an uig/gid leaking issue

alban commented on goaci.go:
```diff
@@ -47,189 +27,401 @@ func (v *StringVector) Set(str string) error {
+				return v, nil
+			}
+		}
+		return "", fmt.Errorf("No such binary found in gobin: %q. There are following binaries available: \"%s\"", opts.useBinary, strings.Join(names, `", "`))
```
This line has both %q and %s. Is it intentional?

krnowak replied:
Urgh, one second. Fix coming.

iaguis commented on goaci.go:
```diff
@@ -47,189 +27,401 @@ func (v *StringVector) Set(str string) error {
+			names = append(names, v.Name())
+		}
+		if opts.useBinary == "" {
+			return "", fmt.Errorf("Found multiple binaries in gobin, but --use-binary option is not used. Please specify which binary to put in ACI. Following binaries are available: \"%s\"", strings.Join(names, `", "`))
```
Missing %q

iaguis commented on goaci.go:
```diff
@@ -47,189 +27,401 @@ func (v *StringVector) Set(str string) error {
+	}
+
+	if opts.keepTmp {
+		Info(`Preserving temporary directory "`, pathsNames.tmpDirPath, `"`)
```
It's annoying that Info et al. don't support %q :stuck_out_tongue_closed_eyes:

alban commented on goaci.go:
```diff
@@ -47,189 +27,401 @@ func (v *StringVector) Set(str string) error {
+
+	// FIXME: the files in the tar archive are added with the
+	// wrong uid/gid. The uid/gid of the aci builder leaks in the
+	// tar archive. See: #16
```
I meant to write the full URL in the comment but github converted my comment in a short link.
`https://github.com/appc/goaci/issues/16`

alban commented:
Except the couple of nitpicks, all the new commits look good to me.

krnowak committed:
Use %q instead of "%s"

iaguis commented:
> Except the couple of nitpicks, all the new commits look good to me.

Ditto.

krnowak committed:
Put an URL to issue instead of just a number
